src/main/java/org/quiltmc/loader/api/QuiltLoader.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2016 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.loader.api;
+
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import org.jetbrains.annotations.Nullable;
+import org.quiltmc.loader.api.entrypoint.EntrypointContainer;
+import org.quiltmc.loader.impl.QuiltLoaderImpl;
+
+import net.fabricmc.loader.api.EntrypointException;
+import net.fabricmc.loader.api.LanguageAdapter;
+
+import net.fabricmc.api.EnvType;
+
+/**
+ * The public-facing FabricLoader instance.
+ *
+ * <p>To obtain a working instance, call {@link #getInstance()}.</p>
+ *
+ * @since 0.4.0
+ */
+public interface QuiltLoader {
```

Comment:
Let's think ahead of time here on plugins:

Do we need to restrict any of the functions here and they use during the plugin loading stage? If yes then the quilt loader instance should be passed via the game entrypoint (and pre launch) and the singleton kept as impl detail.

Reply:
While that's *technically* true (before the game has loaded it doesn't make sense for plugins to call certain methods) I think it would make it much worse from a usage perspective - you can only load one set of mods for minecraft, not two - so it will always make sense for `QuiltLoaderImpl` to only have one instance, and it's a lot easier to program mods if you can reliably get that instance in a static method. *(The alternative is to make all mods provide a `PreLaunchEntrypoint` just to get the QuiltLoader instance as early as possible, which sounds a bit nasty)*

Essentially I'd rather just redefine some of the methods to throw an exception if they are called too early - since there will be a lot more mods developed on quilt than plugins for quilt.